helper/win/drive.h

```diff
@@ -0,0 +1,113 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#ifndef DRIVE_H
+#define DRIVE_H
+
+#include <type_traits>
+#include <stdexcept>
+#include <utility>
+
+#include <QObject>
+#include <QProcess>
+#include <QString>
+#include <QtGlobal>
+
+#include <io.h>
+#include <windows.h>
+
+#include "genericdrive.h"
+
+class Drive : public GenericDrive {
+    Q_OBJECT
+private:
+    static void throwError(const QString &error);
+    static HANDLE openBlockDevice(const QString &device);
+    static char unusedDriveLetter();
+    template <class OutBuffer, class InBuffer = std::nullptr_t>
+    static DWORD controlCodeOf();
+    template <class T>
+    static T *addressOf(T &variable);
+    template <class T>
+    static std::size_t sizeOf();
+    template <class OutBuffer = std::nullptr_t, class InBuffer = std::nullptr_t>
+    static OutBuffer deviceIoControl(HANDLE handle, const QString &message = "Failed", InBuffer *inbuffer = nullptr);
+
+    template <class OutBuffer = std::nullptr_t, class InBuffer = std::nullptr_t>
+    OutBuffer deviceIoControl(const QString &message = "Failed", InBuffer *inbuffer = nullptr) const;
+    bool deviceIoControlCode(DWORD controlCode) const;
+    void lock();
+    void unlock();
+    bool hasDriveLetter(const char driveLetter) const;
+
+public:
+    /**
+     * Shared public interface across platforms.
+     */
+    explicit Drive(const QString &driveIdentifier);
+    ~Drive();
+    void init();
+    void write(const void *buffer, std::size_t size);
+    int getDescriptor() const;
+    void wipe();
+    void umount();
+
+private:
+    HANDLE m_handle;
+    DWORD m_driveNumber;
+    DISK_GEOMETRY_EX m_geometry;
+};
+
+template <class T>
+std::size_t Drive::sizeOf() {
+    if (std::is_same<T, std::nullptr_t>::value)
+        return 0;
+    return sizeof(T);
+}
+
+template <class T>
+T *Drive::addressOf(T &variable) {
+    if (std::is_same<T, std::nullptr_t>::value)
+        return nullptr;
+    return &variable;
+}
+
+/**
+ * The design of the DeviceIoControl is evil partially because it does too many
+ * things.
+ * This wrapper does not fix that but it makes the interface easier to use
+ * since the user does not need to specify the control code or provide unused
+ * arguments when calling the wrapper.
+ */
+template <class OutBuffer, class InBuffer>
```

Comment:
The reason for this is that if I call `DeviceIoControl` directly I need to declare two variables for each call which are not directly initialized (which is evil) and of which one is not used. Then I need to put each call in an if statement for error handling which makes the whole thing a 5 line operation for each call. Now this is a one liner like below which also optionally can handle custom error messages:
```
VOLUME_DISK_EXTENTS vde = deviceIoControl<decltype(vde)>(handle);
```
vs.
```
VOLUME_DISK_EXTENTS vde;
DWORD bytesReturned;
// Way to long line by the way.
if (!::DeviceIoControl(handle, IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS, nullptr, 0, &vde, sizeof(vde), &bytesReturned, nullptr)) {
    throwError("Failed");
}
```
The main reason why I wrote all this is that when I used `::DeviceIoControl` calls even more often since I tried to build the entire thing with `::DeviceIoControl` at one point I always had to lookup the control code everytime or forgot to change the control code when copying it because it's so many lines. So I decided to hard code the type <-> control code stuff once and make this generic.

I'll provide more documentation on how to use and extend this if someone would like to use it in the future.

Reply:
Please provide some brief description in the comments before we merge this. I think it's pretty important to have it.

This is not a competition to write as little code as possible so I'd say either have it documented or don't have it at all if it makes it harder to get the full understanding of how the code actually works.

Personally, I'm not very well versed in writing template code so it's quite challenging for me.